File: n26/format_n26.go
package n26

import (
	"errors"
	"rahul-rakshit/formatbankcsv/output"
	"rahul-rakshit/formatbankcsv/utils"
)

func FormatN26(inputLines [][]string) ([][]string, error) {
	requiredColumns := []string{"Value Date", "Partner Name", "Payment Reference", "Amount (EUR)"}

	header := inputLines[0]
	if !utils.HasColumns(header, requiredColumns) {
		return nil, errors.New("Unexpected header for n26 format")
	}

	columnIndices := make(map[string]int)
	for i, col := range header {
		columnIndices[col] = i
	}

	var requiredIndices []int
	for _, col := range requiredColumns {
		for i, headerCol := range header {
			if headerCol == col {
				requiredIndices = append(requiredIndices, i)
				break
			}
		}
	}

	outputLines := [][]string{
		output.Header,
	}
	for _, row := range inputLines[1:] {
		extractedRow := make([]string, len(requiredIndices))
		for i, index := range requiredIndices {
			extractedRow[i] = row[index]
		}
		outputLines = append(outputLines, extractedRow)
	}

	return outputLines, nil
}

File: n26/formatN26.ts
import { hasColumns } from "../utils/hasColumns.ts";
import { Header } from "../utils/constants.ts";

export function formatN26(inputLines: string[][]): string[][] {
  const requiredColumns = ["Booking Date", "Partner Name", "Payment Reference", "Amount (EUR)"];

  const header = inputLines[0];
  if (!hasColumns(header, requiredColumns)) {
    throw new Error("Unexpected header for n26 format");
  }

  const columnIndices = new Map<string, number>();
  for (let i = 0; i < header.length; i++) {
    columnIndices.set(header[i], i);
  }

  const requiredIndices: number[] = [];
  for (const col of requiredColumns) {
    for (let i = 0; i < header.length; i++) {
      if (header[i] === col) {
        requiredIndices.push(i);
        break;
      }
    }
  }

  const outputLines: string[][] = [[...Header]];
  
  for (const row of inputLines.slice(1)) {
    const extractedRow = requiredIndices.map(index => row[index]);
    outputLines.push(extractedRow);
  }

  return outputLines;
}
